### ISSUE TITLE
[ES Health API] Improvements to the UNKNOWN status

### ISSUE THREAD

**Issue by cristina-eleonora** [elastic/elasticsearch#86924]
## Issue

TLDR; When the status of an issue is unknown, a user doesn't know what's the impact, which indicator reports it, not what to do to fix it.

**Details**

When the master node is unavailable (indicator `instance_has:_master`), it impacts the status of all the health components and part of their indicators.

|Component| Status|
|---|---|
| Cluster coordination | 🔴 Red |
| Data | ⚪ Unknown |
| Snapshots | ⚪ Unknown |

Breakdown by indicator:

| Indicator | Status | Component | Message displayed in the "Summary" |
|---|---|---|---|
|instance_has_master| Red | cluster coordination | Health coordinating instance does not have a master node. |
|ilm| Unknown| data |Could not determine indicator state. Cluster state is not stable. Check details for critical issues keeping this indicator from running |
|has_shards_available| Unknown| data | Could not determine indicator state. Cluster state is not stable. Check details for critical issues keeping this indicator from running |
|slm| Unknown| data | Could not determine indicator state. Cluster state is not stable. Check details for critical issues keeping this indicator from running |
|repository_integrity| Unknown| data | Could not determine indicator state. Cluster state is not stable. Check details for critical issues keeping this indicator from running |

<details>
    <summary>View the API response </summary>

```
> GET _internal/_health
{
  "status": "red",
  "cluster_name": "elasticsearch",
  "components": {
    "cluster_coordination": {
      "status": "red",
      "indicators": {
        "instance_has_master": {
          "status": "red",
          "summary": "Health coordinating instance does not have a master node.",
          "details": {
            "coordinating_node": {
              "node_id": "RwqgEBxlRyOs0uWa12YaFA",
              "name": "kanelbullar"
            },
            "master_node": {
              "node_id": null,
              "name": null
            }
          },
          "impacts": [
            {
              "severity": 1,
              "description": "The cluster cannot create, delete, or rebalance indices, and cannot insert or update documents.",
              "impact_areas": [
                "ingest"
              ]
            },
            {
              "severity": 1,
              "description": "Scheduled tasks such as Watcher, ILM, and SLM will not work. The _cat APIs will not work.",
              "impact_areas": [
                "deployment_management"
              ]
            },
            {
              "severity": 3,
              "description": "Snapshot and restore will not work.",
              "impact_areas": [
                "backup"
              ]
            }
          ]
        }
      }
    },
    "data": {
      "status": "unknown",
      "indicators": {
        "shards_availability": {
          "status": "unknown",
          "summary": "Could not determine indicator state. Cluster state is not stable. Check details for critical issues keeping this indicator from running.",
          "details": {
            "reasons": {
              "instance_has_master": "red"
            }
          }
        },
        "ilm": {
          "status": "unknown",
          "summary": "Could not determine indicator state. Cluster state is not stable. Check details for critical issues keeping this indicator from running.",
          "details": {
            "reasons": {
              "instance_has_master": "red"
            }
          }
        }
      }
    },
    "snapshot": {
      "status": "unknown",
      "indicators": {
        "repository_integrity": {
          "status": "unknown",
          "summary": "Could not determine indicator state. Cluster state is not stable. Check details for critical issues keeping this indicator from running.",
          "details": {
            "reasons": {
              "instance_has_master": "red"
            }
          }
        },
        "slm": {
          "status": "unknown",
          "summary": "Could not determine indicator state. Cluster state is not stable. Check details for critical issues keeping this indicator from running.",
          "details": {
            "reasons": {
              "instance_has_master": "red"
            }
          }
        }
      }
    }
  }
}
```
</details>

Translated to the cloud experience, the UI would look like this:

<img width="1338" alt="Screenshot 2022-05-19 at 13 18 00" src="https://user-images.githubusercontent.com/44808440/169281393-14ba6192-6f81-4bb4-bd2f-1ca03f2ba357.png">

The health issues that have the status `unknown`:
- look like a repetition of each other
- don't mention which indicator is affected
- don't have any impact areas
- reading the description it's not very clear that the critical issue at the top is what's causing the other 4

Since the payload for each indicator is poor in information and **does not have any impacts or user actions**, it does not allow us to populate the issue details page with any useful information.

Details page in the cloud UI

<img width="1269" alt="Screenshot 2022-05-19 at 13 53 26" src="https://user-images.githubusercontent.com/44808440/169287339-5dcae231-a479-4f03-ac2c-7bc51d83a6d2.png">

Example API response:

```
"shards_availability": {
    "status": "unknown",
    "summary": "Could not determine indicator state. Cluster state is not stable. Check details for critical issues keeping this indicator from running.",
        "details": {
            "reasons": {
                 "instance_has_master": "red"
             }
         }
}        
```

## Suggested solution

When the status is **unknown**, let's:
- add the affected indicator name in the summary
- introduce the reason, either in the summary, or in an impact section
    - The API response has a "reasons" field which appears that could let us pinpoint exactly why the status is unknown
- Let's add some impacts and user actions if possible
- Could we add impact areas, or are they difficult to identify?
    - If not, could we just display `Not available`?

This way, the issues page could look like this:

<img width="1201" alt="Screenshot 2022-05-19 at 14 49 14" src="https://user-images.githubusercontent.com/44808440/169297094-1b73564b-83ce-4964-9be2-6fce3383e5ff.png">

And the details page of an unknown issue could look like this:

<img width="1189" alt="Screenshot 2022-05-19 at 14 53 23" src="https://user-images.githubusercontent.com/44808440/169297911-2aa379dc-64c4-44f5-8cde-b10eefdc1848.png">

The API response could look like this:

```
"ilm": {
    "status": "unknown",
    "summary": "Could not determine ILM state.",
    "details": {
        "reasons": {
            "instance_has_master": "red"
        }
    },
    "impacts": [
        {
            "description": "Could not determine ILM state. Cluster state is not stable because the health coordinating instance does not have a master node. ",
            "impact_areas": [
                "Not available" // would be great if we could have real impact areas
            ]
        }
    ],
    "user_actions": [
        {
            "message": "Review and fix all critical health issues that affect master nodes.",
        }
    ]    
}
```

CC: @andreidan @Leaf-Lin @shubhaat @elastichelix @cjcenizal

**Comment by andreidan** [comment]
@cristina-eleonora Thanks for filing this. 

Every indicator with the YELLOW or GREEN status will eventually offer impacts and user actions. 

If the `instance_has_master` indicator is RED, can you maybe provide more details as to what's the user value from seeing several `UNKOWN` statuses? 
Would only showing the RED/YELLOW/GREEN indicators be enough?  ie. there's one actionable issue already (contact support to fix the cluster health via the `instance_has_master` indicator) 

The API can return UNKNOWN in a couple of scenarios:
1. All indicators are UNKOWN because the coordinating node that receives the health request didn’t yet recover its local state (to be able to know about things in the cluster). The user should try the API on a different node, or try again later on the same node.	
The `summary` field in this case indicates this (e.g. `Try again later or execute the health API against a different node.`) which IMO is what we should display when the status is `UNKOWN`. That's if we even want to surface this - I also imagine we could perform this retry for the user ourselves? 

2.  `instance_has_master` (or soon the `master_is_stable`) indicator reported RED and the other indicators are reporting UNKOWN. I believe we can in this case pull the cause of why the indicators are `UNKNOWN` from the details section into the `summary` if that's better/easier (//cc @jbaiera ).
However, it’s not possible for us to accurately offer any advice or indication of impacts. A user must attend to the RED indicator to get back to a place where we can offer more data.

I believe if we lift the indicator that caused the UNKNOWN status into the `summary` we will have addressed your first two suggestions (namely: `add the affected indicator name in the summary` and `introduce the reason, either in the summary, or in an impact section`) 

On your remaining suggestions
 * `Let's add some impacts and user actions if possible`
 *  `Could we add impact areas, or are they difficult to identify?`

IMO we should not do this because we don't know and shouldn't speculate. The `summary` field will offer a suggestion in the 1st scenario described above (ie. retry the api call on a different node) and for the 2nd scenario above the user will be able to act on the `instance_has_master` indicator to get it to `GREEN`.

// @Leaf-Lin

**Comment by cristina-eleonora** [comment]
Thank you Andrei for the thorough answer 💟  

> 1. All indicators are UNKOWN because the coordinating node that receives the health request didn’t yet recover its local state (to be able to know about things in the cluster). The user should try the API on a different node, or try again later on the same node.

Could we formulate the text to say that the status could not be retrieved because the Elasticsearch node responsible of the health check is not responsive/available. And offer as actions: 
1. Wait x minutes and refresh the page
2. Check the status using another Elasticsearch node

Cloud users wouldn't necessarily know how to perform the 2nd option, but maybe we can find a workaround. Could the API identify if there are other Elasticsearch nodes that have the local state? If yes, perhaps we can use them to run the ES health API and get an accurate state.  Cc @elastichelix @shubhaat

>    The `summary` field in this case indicates this (e.g. `Try again later or execute the health API against a different node.`) which IMO is what we should display when the status is `UNKOWN`. That's if we even want to surface this - I also imagine we could perform this retry for the user ourselves?

++ on performing the retry automatically


>    However, it’s not possible for us to accurately offer any advice or indication of impacts. A user must attend to the RED indicator to get back to a place where we can offer more data.

If we have the reason, could we mention it while we tell the user to solve any critical issues (the RED indicator)? It would be very helpful, users would know what to search for. 


> * `Let's add some impacts and user actions if possible`
> * `Could we add impact areas, or are they difficult to identify?`
> 
> IMO we should not do this because we don't know and shouldn't speculate. The `summary` field will offer a suggestion in the 1st scenario described above (ie. retry the api call on a different node) and for the 2nd scenario above the user will be able to act on the `instance_has_master` indicator to get it to `GREEN`.

Fair enough 🙂

**Comment by DaveCTurner** [comment]
> All indicators are UNKNOWN because the coordinating node that receives the health request didn’t yet recover its local state (to be able to know about things in the cluster).

We discussed this in a separate channel and concluded that the coordinating node will be always able to run the `instance_has_master` indicator at least so there'll never be a case where all indicators report `UNKNOWN`.

**Comment by shubhaat** [comment]
Thansk @DaveCTurner, @andreidan and @cristina-eleonora. 

Based on the latest comment, if there is never a case all indicators will return `UNKNOWN` is this something we don't need to worry about?

**Comment by andreidan** [comment]
>Based on the latest comment, if there is never a case all indicators will return UNKNOWN is this 
something we don't need to worry about?

@shubhaat Yes, that's correct 🙌
